cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/gvn/ValueNumbering.qll

```diff
@@ -85,236 +61,39 @@ class ValueNumber extends TValueNumber {
-) {
-  instr.getEnclosingIRFunction() = irFunc and
-  instr.getResultIRType() = type and
-  instr.getValue() = value
-}
-
-private predicate stringConstantValueNumber(
-  StringConstantInstruction instr, IRFunction irFunc, IRType type, string value
-) {
-  instr.getEnclosingIRFunction() = irFunc and
-  instr.getResultIRType() = type and
-  instr.getValue().getValue() = value
-}
-
-private predicate fieldAddressValueNumber(
-  FieldAddressInstruction instr, IRFunction irFunc, Language::Field field, ValueNumber objectAddress
-) {
-  instr.getEnclosingIRFunction() = irFunc and
-  instr.getField() = field and
-  valueNumber(instr.getObjectAddress()) = objectAddress
-}
 
-private predicate binaryValueNumber(
-  BinaryInstruction instr, IRFunction irFunc, Opcode opcode, IRType type, ValueNumber leftOperand,
-  ValueNumber rightOperand
-) {
-  instr.getEnclosingIRFunction() = irFunc and
-  not instr instanceof PointerArithmeticInstruction and
-  instr.getOpcode() = opcode and
-  instr.getResultIRType() = type and
-  valueNumber(instr.getLeft()) = leftOperand and
-  valueNumber(instr.getRight()) = rightOperand
-}
-
-private predicate pointerArithmeticValueNumber(
-  PointerArithmeticInstruction instr, IRFunction irFunc, Opcode opcode, IRType type,
-  int elementSize, ValueNumber leftOperand, ValueNumber rightOperand
-) {
-  instr.getEnclosingIRFunction() = irFunc and
-  instr.getOpcode() = opcode and
-  instr.getResultIRType() = type and
-  instr.getElementSize() = elementSize and
-  valueNumber(instr.getLeft()) = leftOperand and
-  valueNumber(instr.getRight()) = rightOperand
-}
-
-private predicate unaryValueNumber(
-  UnaryInstruction instr, IRFunction irFunc, Opcode opcode, IRType type, ValueNumber operand
-) {
-  instr.getEnclosingIRFunction() = irFunc and
-  not instr instanceof InheritanceConversionInstruction and
-  not instr instanceof CopyInstruction and
-  instr.getOpcode() = opcode and
-  instr.getResultIRType() = type and
-  valueNumber(instr.getUnary()) = operand
-}
-
-private predicate inheritanceConversionValueNumber(
-  InheritanceConversionInstruction instr, IRFunction irFunc, Opcode opcode,
-  Language::Class baseClass, Language::Class derivedClass, ValueNumber operand
-) {
-  instr.getEnclosingIRFunction() = irFunc and
-  instr.getOpcode() = opcode and
-  instr.getBaseClass() = baseClass and
-  instr.getDerivedClass() = derivedClass and
-  valueNumber(instr.getUnary()) = operand
-}
-
-/**
- * Holds if `instr` should be assigned a unique value number because this library does not know how
- * to determine if two instances of that instruction are equivalent.
- */
-private predicate uniqueValueNumber(Instruction instr, IRFunction irFunc) {
-  instr.getEnclosingIRFunction() = irFunc and
-  not instr.getResultIRType() instanceof IRVoidType and
-  not numberableInstruction(instr)
+  Language::Expr getAnExpr() { result = getAnInstruction().getUnconvertedResultExpression() }
```

Comment:
I can believe that using the unconverted result will make the wrapper's results more similar to the AST-based implementation, but won't they also be less correct? If so, let's add a second member predicate, maybe named `getConvertedExpr`, that might return a `Conversion` if that's where the value number is best placed. There would also need to be a sibling predicate for `valueNumber`.

This is similar to [what we have in IR data flow](https://github.com/Semmle/ql/blob/931c0e982e52e6b848a8c46704d37b09a7be686b/cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll#L41-L56). On a related note, couldn't the implementation of `Node::asExpr` be replaced with just `result = instr.getUnconvertedResultExpression()`? Its implementation is [here](https://github.com/Semmle/ql/blob/master/cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/IRConstruction.qll#L69-L79).

Reply:
That's a good point!
Given that the predicate for obtaining an unconverted expression is called `getAnExpr`, the best name I can think of is to call the new predicate `getAConvertedExpr`? Should we then add an alias for `getAnExpr` called `getAnUnconvertedExpr` to make the choice of unconverted expression explicit when using the library?

Reply:
Yes and yes.